module/zfs/abd.c

```diff
@@ -346,8 +334,44 @@ abd_alloc_pages(abd_t *abd, size_t size)
 		}
 	}
 
-	ABD_SCATTER(abd).abd_sgl = table.sgl;
-	ABD_SCATTER(abd).abd_nents = table.nents;
+	/*
+	 * These conditions ensure that the transformation to a linear ABD
+	 * is valid.
+	 */
+	ASSERT(!PageHighMem(sg_page(ABD_SCATTER(abd).abd_sgl)));
+	ASSERT0(ABD_SCATTER(abd).abd_offset);
+
+	if (table.nents == 1) {
```

Comment:
Since `chunks` must equal `table.nents` I think it would be helpful if we updated the `if (chunks > 1)` check above to be `table.nents > 0`.  That would make it clear these cases are mutually exclusive.  More over it looks like that block could be safely moved in to an `else if` block in this `if` conditional.  Maybe a little restructuring is in order, something like this?

```c
if (table.nents == 1)  {
	...
} else if (table.nents > 1) {
	...
	ABD_SCATTER(abd).abd_sgl = table.sgl;
	ABD_SCATTER(abd).abd_nents = table.nents;
}
```

Reply:
Good idea, will do.

This also raises the question of if nents can really be 0.  Unfortunately it looks like the abd code does not enforce that the size / nents is nonzero, so I will continue assuming that it could be 0 (although I doubt this case is useful) and structure the code as you mentioned.